core/src/main/java/io/ddf/DDFCache.java

```diff
@@ -2,52 +2,166 @@
 
 
 import com.google.common.base.Strings;
+import com.google.common.cache.*;
+import com.google.common.util.concurrent.Striped;
 import io.ddf.exception.DDFException;
 
+import java.util.ArrayList;
+import java.util.List;
 import java.util.Map;
 import java.util.UUID;
 import java.util.concurrent.ConcurrentHashMap;
 
+import io.ddf.misc.ALoggable;
+import io.ddf.misc.Config;
+import org.apache.commons.lang.exception.ExceptionUtils;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+
 /**
- * Created by huandao on 6/11/15.
  */
-public class DDFCache {
+public class DDFCache extends ALoggable {
+
+  Striped<Lock> uuidLocks = Striped.lock(Runtime.getRuntime().availableProcessors() * 4);
+
+  /**
+   * Keeping DDF without name
+   */
+  private Cache<UUID, DDF> mDDFCache;
+
+  /**
+   * Keeping DDF that has name in a separate cache, so eviction on DDF without name
+   * does not affect DDF with name
+   */
+  private Cache<UUID, DDF> mDDFwithNameCache;
+
+  private DDFManager mDDFManager;
+
+  public DDFCache(DDFManager manager) {
+
+    mDDFManager = manager;
+
+    /**
+     * Maximum number of DDFs without name allowed in cache, DDFs will be evicted by an approximate LRU algorithm
+     */
+    Long maxNumberOfDDFs = Long.valueOf(Config.getGlobalValue(Config.ConfigConstant.MAX_NUMBER_OF_DDFS_IN_CACHE));
+
+    /**
+     * Maximum number of DDFs with name allowed in cache, DDFs with name will be evicted by an approximate LRU algorithm
+     */
+    Long maxNumberOfDDFsWName = Long.valueOf(Config.getGlobalValue(Config.ConfigConstant.MAX_NUMBER_OF_DDFS_WITH_NAME_IN_CACHE));
+
+    /**
+     * The number of seconds that the DDF will live in cache since it last used(meaning inserted or accessed)
+     */
+    Long timeToIdle = Long.valueOf(Config.getGlobalValue(Config.ConfigConstant.DDF_TIME_TO_IDLE_SECONDS));
+
+    mLog.info(String.format("Maximum number of ddfs in cache %s", maxNumberOfDDFs));
+    mLog.info(String.format("DDF's time to idle = %s seconds", timeToIdle));
+
+    mDDFCache = CacheBuilder.newBuilder().
+        maximumSize(maxNumberOfDDFs).expireAfterAccess(timeToIdle, TimeUnit.SECONDS).
+        recordStats().removalListener(new DDFRemovalListener())
+        .build();
+
+    // Don't do time based eviction for DDF with name
+    mDDFwithNameCache = CacheBuilder.newBuilder().
+        maximumSize(maxNumberOfDDFsWName).recordStats().removalListener(new DDFRemovalListener())
+        .build();
+  }
+
+  private boolean containsDDF(UUID uuid) {
+    return (mDDFCache.asMap().containsKey(uuid) || mDDFwithNameCache.asMap().containsKey(uuid));
+  }
 
-  private Map<UUID, DDF> mDDFs = new ConcurrentHashMap<UUID, DDF>();
   private Map<String, UUID> mUris = new ConcurrentHashMap<String, UUID>();
 
   public void addDDF(DDF ddf) throws DDFException {
```

Comment:
Your idea is locking on uuid of ddf which could change wherever on user code space so it doesn't (or couldn't) lock here.